bosh-deployment is not a valid label name, prometheus metrics

## Summary

Upgrading to Concourse v7.4.0, I receive the following errors on our web nodes:

```
[...]
panic: descriptor Desc{fqName: "concourse_error_logs", help: "Number of error logged", constLabels: {}, variableLabels: [message]} is invalid: "bosh-deployment" is not a valid label name for metric "concourse_error_logs"

goroutine 1 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc0000a4ff0, 0xc0007b28b0, 0x1, 0x1)
        github.com/prometheus/client_golang@v1.11.0/prometheus/registry.go:403 +0xb7
github.com/prometheus/client_golang/prometheus.MustRegister(...)
        github.com/prometheus/client_golang@v1.11.0/prometheus/registry.go:178
github.com/concourse/concourse/atc/metric/emitter.(*PrometheusConfig).NewEmitter(0xc000518ce0, 0xc000779b60, 0x0, 0x0, 0x1, 0xc0007b2850)
        github.com/concourse/concourse/atc/metric/emitter/prometheus.go:133 +0x184
github.com/concourse/concourse/atc/metric.(*Monitor).Initialize(0xc00009e000, 0x36cca90, 0xc00053dc80, 0xc0007a26f0, 0x24, 0xc000779b60, 0x3e8, 0xc00053dc80, 0xc000320340)
        github.com/concourse/concourse/atc/metric/emit.go:143 +0x4c5
github.com/concourse/concourse/atc/atccmd.(*RunCommand).configureMetrics(0xc0001d8b00, 0x36cca90, 0xc00053db00, 0x36614c0, 0x335b5d8)
        github.com/concourse/concourse/atc/atccmd/command.go:1599 +0xd5
github.com/concourse/concourse/atc/atccmd.(*RunCommand).Runner(0xc0001d8b00, 0xc00079be20, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0)
        github.com/concourse/concourse/atc/atccmd/command.go:571 +0x545
main.(*WebCommand).Runner(0xc0002d4008, 0xc00079be20, 0x0, 0x1, 0x2, 0x6, 0xc00079bc90, 0xc00028fd00)
        github.com/concourse/concourse/cmd/concourse/web.go:62 +0xa5
main.(*WebCommand).Execute(0xc0002d4008, 0xc00079be20, 0x0, 0x1, 0x2778480, 0x2af9320)
        github.com/concourse/concourse/cmd/concourse/web.go:44 +0x65
github.com/vito/twentythousandtonnesofcrudeoil.installEnv.func2(0x7fa6563cdcc0, 0xc0002d4008, 0xc00079be20, 0x0, 0x1, 0x1, 0x182)
        github.com/vito/twentythousandtonnesofcrudeoil@v0.0.0-20180305154709-3b21ad808fcb/environment.go:40 +0x8a
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc00027d880, 0xc000130010, 0x1, 0x1, 0xc00015e640, 0x0, 0x0, 0x0, 0x0)
        github.com/jessevdk/go-flags@v1.4.1-0.20200711081900-c17162fe8fd7/parser.go:340 +0x85d
github.com/jessevdk/go-flags.(*Parser).Parse(...)
        github.com/jessevdk/go-flags@v1.4.1-0.20200711081900-c17162fe8fd7/parser.go:190
main.main()
        github.com/concourse/concourse/cmd/concourse/main.go:30 +0x1f5
```

Looks like it's trying to specify an invalid label, I believe dashes aren't allowed in label names for Prometheus.

## Steps to reproduce

Using this bosh config when deploying:

```yaml
name: concourse
instance_groups:
[...]
- azs:
  - AZ1
  instances: 4
  jobs:
  - name: web
    properties:
      add_local_users: [...]
      audit:
        build: true
        job: true
        pipeline: true
        system: true
        team: true
      auth_duration: 48h
      bind_ip: 127.0.0.1
      bind_port: 8080
      build_log_retention:
        default_builds: 20
        default_days: 30
        maximum_builds: 50
        maximum_days: 60
      build_tracker_interval: 10s
      capture_error_metrics: true
      container_placement_strategies:
      - limit-active-tasks
      - volume-locality
      - fewest-build-containers
      default_check_interval: 2m
      default_resource_type_check_interval: 30m
      default_task_cpu_limit: 150
      default_task_memory_limit: 4096mb
      enable_archive_pipeline: true
      enable_cache_streamed_volumes: true
      enable_global_resources: false
      enable_pipeline_instances: true
      enable_rerun_when_worker_disappears: true
      enable_skip_checking_not_in_use_resources: true
      external_url: https://[...]
      generic_oidc:
        client_id: [...]
        client_secret: [...]
        display_name: Single SignOn
        groups_key: groups
        issuer: [...]
        scopes: https://graph.microsoft.com/.default profile email openid
        user_name_key: email
      global_resource_check_timeout: 10m
      job_scheduling_max_in_flight: 8
      ldap_auth:
        bind_dn: [...]
        bind_pw: [...]
        display_name: [...]
        group_search_base_dn: [...]
        group_search_filter: [...]
        group_search_group_attr: member
        group_search_name_attr: cn
        group_search_scope: sub
        group_search_user_attr: distinguishedName
        host: [...]
        insecure_skip_verify: true
        start_tls: false
        user_search_base_dn: [...]
        user_search_email_attr: mail
        user_search_filter: (objectClass=person)
        user_search_id_attr: sAMAccountName
        user_search_name_attr: displayName
        user_search_scope: sub
        user_search_username: sAMAccountName
      lidar_scanner_interval: 2m
      main_team:
        auth:
          ldap:
            groups: [...]
            users: []
          local:
            users: [...]
          oidc:
            groups: [...]
      max_active_tasks_per_worker: 10
      p2p:
        enable: true
        stream_timeout: 15m
      postgresql:
        database: atc
        role:
          name: [...]
          password: [...]
      prometheus:
        bind_ip: 0.0.0.0
        bind_port: 8456
      redact_secrets: false
      secrets:
        cache:
          duration: 2m
          duration_notfound: 2m
          enabled: true
          purge_interval: 10m
        retry_attempts: 3
        retry_interval: 30s
      streaming_artifacts_compression: gzip
      token_signing_key:
        private_key: [...]
        public_key: [...]
      vault:
        auth:
          client_token: [...]
        path_prefix: /concourse
        query_timeout: 30s
        shared_path: /concourse-template
        url: [...]
      worker_gateway:
        authorized_keys: [...]
        host_key:
          private_key: [...]
          public_key: [...]
    release: concourse
  name: web
  networks:
  - default:
    - dns
    - gateway
    name: default
    static_ips: [...]
  stemcell: default
  vm_type: cc-app-web
releases:
- name: "concourse"
  version: "7.4.0"
  url: "https://bosh.io/d/github.com/concourse/concourse-bosh-release?v=7.4.0"
  sha1: "a0737a17fc46490ede1e83516f02a92c009417a6"
- name: bpm
  sha1: 502e9446fa34accaf122ad2b28b6ffa543d5bbca
  url: https://bosh.io/d/github.com/cloudfoundry/bpm-release?v=1.1.12
  version: 1.1.12
- name: postgres
  sha1: e44bbe8f8a7cdde1cda67b202e399a239d104db6
  url: https://bosh.io/d/github.com/cloudfoundry/postgres-release?v=43
  version: "43"
- name: syslog
  sha1: 3c6636831f3c89fa9e8f4b34613b5cefd3b29046
  url: https://bosh.io/d/github.com/cloudfoundry/syslog-release?v=11.7.5
  version: 11.7.5
- name: nginx
  sha1: 7850fda50d355316c49d16a4e1c4d30227204733
  url: https://bosh.io/d/github.com/cloudfoundry-community/nginx-release?v=1.19.10
  version: 1.19.10
- name: os-conf
  sha1: 4f653168954749992a541d228dd4f936f2eff2d6
  url: https://bosh.io/d/github.com/cloudfoundry/os-conf-release?v=22.1.1
  version: 22.1.1
- name: cron
  sha1: f0a048b04fa8f064a6170b34741439f91e7838b5
  url: https://bosh.io/d/github.com/cloudfoundry-community/cron-boshrelease?v=1.3.1
  version: 1.3.1
- name: node-exporter
  sha1: cea1edc938873b7204b889173002e0e9fa813e54
  url: https://bosh.io/d/github.com/cloudfoundry-community/node-exporter-boshrelease?v=5.0.0
  version: 5.0.0
- name: bind9
  sha1: 8a6a99046ea1adba07e0a95d600e7f608f5bb305
  url: https://bosh.io/d/github.com/cloudfoundry-community/bind9-boshrelease?v=1.3.0
  version: 1.3.0
```

## Expected results

Expect it to deploy successfully, while also exporting metrics with the bosh deployment being identified.

## Actual results

Due to the error_metrics being enabled, and it trying to specify the "bosh-deployment" label, which isn't a valid Prometheus label, it fails.

## Triaging info

* Concourse version: v7.4.0 via Bosh
* Did this used to work? Yes, however I don't think bosh-deployments was an exported label in previous versions.

--- a/atc/metric/emitter/prometheus.go
+++ b/atc/metric/emitter/prometheus.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"net"
 	"net/http"
+	"regexp"
 	"sort"
 	"strings"
 	"sync"
@@ -107,6 +108,34 @@ func serializeLabels(labels *prometheus.Labels) string {
 	return key
 }
 
+// rePrometheusLabelInvalid matches any invalid characters we may have in our
+// emitted labels.
+//
+// Prometheus format:
+//   > Label names may contain ASCII letters, numbers, as well as underscores.
+//   > They must match the regex [a-zA-Z_][a-zA-Z0-9_]*. Label names beginning
+//   > with __ are reserved for internal use.
+//   Link: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+var rePrometheusLabelInvalid = regexp.MustCompile(`[^a-zA-Z0-9_]+`)
+
+// rePrometheusLabelClean ensures we clean any duplicate underscores.
+var rePrometheusLabelClean = regexp.MustCompile(`__+`)
+
+// sanitizePrometheusLabels ensures all metric labels we register with the prometheus
+// emitter are sanitized to support Prometheus format. See rePrometheusLabelInvalid
+// for more information.
+func sanitizePrometheusLabels(labels map[string]string) map[string]string {
+	newLabels := make(map[string]string)
+
+	for k, v := range labels {
+		k = rePrometheusLabelInvalid.ReplaceAllString(k, "_")
+		k = strings.Trim(rePrometheusLabelClean.ReplaceAllString(k, "_"), "_")
+		newLabels[k] = v
+	}
+
+	return newLabels
+}
+
 func init() {
 	metric.Metrics.RegisterEmitter(&PrometheusConfig{})
 }
@@ -120,6 +149,9 @@ func (config *PrometheusConfig) bind() string {
 }
 
 func (config *PrometheusConfig) NewEmitter(attributes map[string]string) (metric.Emitter, error) {
+	// ensure there are no invalid characters in label names.
+	attributes = sanitizePrometheusLabels(attributes)
+
 	// error log metrics
 	errorLogs := prometheus.NewCounterVec(
 		prometheus.CounterOpts{
@@ -601,8 +633,10 @@ func (config *PrometheusConfig) NewEmitter(attributes map[string]string) (metric
 // Event types (differentiated by the less-than-ideal string Name field) into different
 // Prometheus metrics.
 func (emitter *PrometheusEmitter) Emit(logger lager.Logger, event metric.Event) {
+	// ensure there are no invalid characters in label names.
+	event.Attributes = sanitizePrometheusLabels(event.Attributes)
 
-	//update last seen counters, used to gc stale timeseries
+	// update last seen counters, used to gc stale timeseries
 	emitter.updateLastSeen(event)
 
 	switch event.Name {

--- a/atc/metric/emitter/prometheus.go
+++ b/atc/metric/emitter/prometheus.go
@@ -704,7 +704,7 @@ func (emitter *PrometheusEmitter) errorLogsMetric(logger lager.Logger, event met
 	message, exists := event.Attributes["message"]
 	if !exists {
 		logger.Error("failed-to-find-message-in-event",
-			fmt.Errorf("expected team_name to exist in event.Attributes"))
+			fmt.Errorf("expected message to exist in event.Attributes"))
 		return
 	}
 

--- a/atc/metric/emitter/prometheus_test.go
+++ b/atc/metric/emitter/prometheus_test.go
@@ -188,7 +188,12 @@ var _ = Describe("PrometheusEmitter", func() {
 	})
 
 	JustBeforeEach(func() {
-		prometheusEmitter, err = prometheusConfig.NewEmitter(nil)
+		prometheusEmitter, err = prometheusConfig.NewEmitter(map[string]string{
+			// Ensure invalid labels are sanitized.
+			"invalid-label":     "foo",
+			"__prefix__test__":  "bar",
+			"_prefix_testtwo__": "baz",
+		})
 	})
 
 	It("emits step waiting metric", func() {
@@ -208,7 +213,7 @@ var _ = Describe("PrometheusEmitter", func() {
 		body, _ := ioutil.ReadAll(res.Body)
 		defer res.Body.Close()
 		Expect(res.StatusCode).To(Equal(http.StatusOK))
-		Expect(string(body)).To(ContainSubstring("concourse_steps_waiting{platform=\"darwin\",teamId=\"42\",teamName=\"teamdev\",type=\"get\",workerTags=\"tester\"} 4"))
+		Expect(string(body)).To(ContainSubstring("concourse_steps_waiting{invalid_label=\"foo\",platform=\"darwin\",prefix_test=\"bar\",prefix_testtwo=\"baz\",teamId=\"42\",teamName=\"teamdev\",type=\"get\",workerTags=\"tester\"} 4"))
 		Expect(err).To(BeNil())
 	})
 })

